wetterdienst/provider/dwd/derived/api.py

```diff
@@ -0,0 +1,327 @@
+                "value": float(df.select(parameter.name).item()),
+                "quality": None,
+            }
+        )
+
+    def _collect_station_parameter_or_dataset(
+        self,
+        station_id: str,
+        parameter_or_dataset: ParameterModel,
+    ) -> pl.DataFrame:
+        data = []
+
+        full_date_range_to_fetch = self._get_first_day_of_months_to_fetch(
+            parameter_or_dataset,
+        )
+
+        for period in self.sr.stations.periods:
+            if period not in parameter_or_dataset.dataset.periods:
+                log.info(f"Skipping period {period} for {parameter_or_dataset.name}.")
+                continue
+            date_range_to_fetch_for_period = self._filter_date_range_for_period(
+                date_range=full_date_range_to_fetch,
+                period=period,
+                parameter=parameter_or_dataset,
+            )
+            for first_day_of_month_to_fetch in date_range_to_fetch_for_period:
+                for dataset, _ in groupby(self.sr.stations.parameters, key=lambda x: x.dataset):
+                    url = self._get_values_url(
+                        product_name=dataset.name,
+                        period=period,
+                        start_date=first_day_of_month_to_fetch.strftime("%Y%m"),
+                    )
+                    file = download_file(
+                        url=url,
+                        cache_dir=self.sr.stations.settings.cache_dir,
+                        ttl=CacheExpiry.FIVE_MINUTES,
+                        client_kwargs=self.sr.settings.fsspec_client_kwargs,
+                        cache_disable=self.sr.settings.cache_disable,
+                    )
+
+                    if file.status == 404:
+                        logging.warning(
+                            f"File {url.rsplit('/', 1)[1]} for "
+                            f"station {station_id} not found on server {url}. Skipping."
+                        )
+                        continue
+                    file.raise_if_exception()
+
+                    df = _get_data_from_file(file)
+                    df = self._get_values_for_single_station(
+                        df=df,
+                        station_id=station_id,
+                    )
+
+                    if df is None:
+                        logging.warning(
+                            f"No data found for ID {station_id} at {first_day_of_month_to_fetch.strftime('%m/%Y')}"
+                        )
+                        continue
+
+                    df = self._process_dataframe_to_expected_format(
+                        df=df,
+                        column_name_mapping=self._column_name_mapping,
+                        date=first_day_of_month_to_fetch,
+                        parameter=parameter_or_dataset,
+                    )
+                    data.append(df)
+
+        if len(data) != 0:
+            return pl.concat(data)
+        return pl.DataFrame()
```

Comment:
I'd personally prefer early exiting if no data was collected and return existing data afterwards. This would leave space to do additional operations on the concatenated dataframe afterwards.

```suggestion
        if len(data) == 0:
            return pl.DataFrame()
        return pl.concat(data)
```